### PR TITLE
fix: njit fallback for broken runtime with torch 

### LINKED ISSUE #3507: segmentation fault (SIGSEGV) when computing neighbors
### Please make sure these conditions are met

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of scanpy.
- [ ] (optional) I have confirmed this bug exists on the main branch of scanpy.

### What happened?

Issue in computing neighbors. Likely env dependent, but given env is fresh I was hoping you may provide some guidance

Fresh env:
```
python==3.10
numba==0.61.0
numpy==1.26.4
scanpy==1.11.0
scipy==1.15.2
# using CONDA_SUBDIR=osx-64 for mamba env creation on M3 chip
# Other packages omitted
```

Try running neighbors:

```
sc.pp.neighbors(emb,n_pcs=30)
/Users/karinhrovatin/miniforge3/envs/protein/lib/python3.10/site-packages/threadpoolctl.py:1226: RuntimeWarning: 
Found Intel OpenMP ('libiomp') and LLVM OpenMP ('libomp') loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md
  warnings.warn(msg, RuntimeWarning)
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)

 ```
The packages libiomp and libomp are not within conda env (at least not seen via pip freeze).
Already tried `export OMP_NUM_THREADS=1` that was mentioned in another issue.

### Minimal code sample

```python
see above
```

### Error output

```pytb

```

### Versions

See above

### LINKED ISSUE #4026: segmentation fault in normalize_total (Numba omp) on Apple Silicon
### Please make sure these conditions are met

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of scanpy.
- [ ] (optional) I have confirmed this bug exists on the main branch of scanpy.

### What happened?

Probably related to #3507. Posting it again since I could not find an existing solution. The workaround is to manually set `NUMBA_THREADING_LAYER=workqueue`.

**Problem:** When scanpy and torch are both loaded, `sc.pp.normalize_total` can crash (segmentation fault) with warning
```text
 OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
```
```python
import scanpy as sc
import torch
from scipy.sparse import csr_matrix
X = csr_matrix(torch.eye(100).numpy())
ad = sc.AnnData(X=X)
sc.pp.normalize_total(ad, target_sum=1e4)
```

Densifying `ad.X` first before running `sc.pp.normalize_total` can rescue the crash. Also, importing numpy only will throw the same OMP message but does not lead to segfault crash.

**Root cause:**  Numba's `@njit` with `numba.prange` (parallel execution) in `_normalize_csr`. Numba's default threading layer on macOS/M-chip is `omp` (LLVM OpenMP). On Apple Silicon, Numba 0.64.0 with OpenMP threading throws the OMP warning (omp_set_nested deprecated). When torch is also imported, this somehow leads to crashes with a segfault. 

A potential issue may be `libomp.dylib`. sklearn bundles libomp.dylib with install name `/DLC/sklearn/.dylibs/libomp.dylib`, while torch ships its own libomp.dylib with install name `/opt/llvm-openmp/lib/libomp.dylib`. Both may get loaded as separate OMP runtimes, and the LLVM OMP runtime crashes on detecting a duplicate.

**Potential fix:** Set `NUMBA_THREADING_LAYER=workqueue` which uses Numba's own platform-agnostic thread pool (no OpenMP dependency)

```bash
export NUMBA_THREADING_LAYER=workqueue
```

Or alternatively,

```python
import os; os.environ['NUMBA_THREADING_LAYER'] = 'workqueue'
import scanpy as sc
import torch
```

### Minimal code sample

```python
conda create -n scanpy_test python=3.12 r-base
conda activate scanpy_test
pip install scanpy==1.12 torch==2.11
```

The `r-base` installation somehow is required to replicate the behavior.

```
import scanpy as sc
import torch
from scipy.sparse import csr_matrix
X = csr_matrix(torch.eye(100).numpy())
ad = sc.AnnData(X=X)
sc.pp.normalize_total(ad, target_sum=1e4)
```

### Error output

```pytb
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
[1]    72347 segmentation fault  python
```

### Versions

<details>

```
scanpy	1.12
----	----
packaging	26.0
donfig	0.8.1.post1
typing_extensions	4.15.0
joblib	1.5.3
anndata	0.12.10
kiwisolver	1.5.0
h5py	3.16.0
numpy	2.4.4
PyYAML	6.0.3
session-info2	0.4
natsort	8.4.0
cycler	0.12.1
legacy-api-wrap	1.5
six	1.17.0
threadpoolctl	3.6.0
fast-array-utils	1.4
scipy	1.17.1
python-dateutil	2.9.0.post0
pillow	12.1.1
numba	0.64.0
numcodecs	0.16.5
scikit-learn	1.8.0
setuptools	81.0.0
fsspec	2026.3.0
llvmlite	0.46.0
pandas	2.3.3
matplotlib	3.10.8
google-crc32c	1.8.0
pyparsing	3.3.2
pytz	2026.1.post1
zarr	3.1.6
----	----
Python	3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 17:06:14) [Clang 19.1.7 ]
OS	macOS-26.4-arm64-arm-64bit
CPU	8 logical CPU cores, arm
GPU	No GPU found
Updated	2026-04-01 02:04
```

</details>

### PR DESCRIPTION
**Pull request by JhonatanFelix**
This PR adds a narrow workaround in `fast-array-utils`’ custom `njit` wrapper for the Apple Silicon crash reported in scanpy when `torch` is loaded and a numba parallel path is used. The goal is the same “workaround” path suggested by `flying-sheep` in the[ Scanpy discussion](https://github.com/scverse/scanpy/pull/4027#issuecomment-4175054956): handle this in the shared runtime-dispatch layer, instead of removing parallelism in Scanpy itself.

The approach keeps the normal parallel path by default. But, on the macOS arm64, when `torch` is already loaded and the current threading config is not expicitly pinned to a safe layer like `workqueue` or `tbb`, the wrapper runs a small cached subprocess probe before using the parallel implementation. That probe  will mirror the current Numba threading config, and it will reproduce the relevant import context, and finally checking whether a tiny `@numba.njit(parallel=True)` function actually ran successfully or not. If it does, the parallel version is used as usual. If it fails, the wrapper falls back to the already-compiled serial version and emits a warning.

I also narrowed the probe context baesd on reproduction work in the failing environnment. In that setup, `torch` was the relevant one, so the probe now mirrors only the loaded `torch` state, and the cache key includes that state as well. This keeps the workaround smaller and avoids reusing a cached “safe” result after `torch` is imported later.

Tests cover the probe gating logic, lazy detection behavior, env/config mirroring for the subprocess, cache-key behavior, subprocess success and failure cases, wrapper dispatch between serial and parallel implementations, and correctness of the serial fallback path.

### CODE REVIEW

File: src/fast_array_utils/numba.py
Comment (flying-sheep) [inline src/fast_array_utils/numba.py:1]: Please rename `numba.py` to `numba/__init__.py` and move everything that has to do with this new workaround into a new submodule of that.

you can even `import numba` at the toplevel there, as you only need to import `_needs_parallel_runtime_probe` and` _parallel_numba_runtime_is_safe` at runtime in the decorator.

File: src/fast_array_utils/numba.py [hunk 1]
```diff
@@ -84,6 +93,108 @@ def _is_in_unsafe_thread_pool() -> bool:
...
 
+def _is_apple_silicon() -> bool:
+    return sys.platform == "darwin" and platform.machine() == "arm64"
+
+
+def _is_torch_loaded() -> bool:
+    return "torch" in sys.modules
+
+
+def _configured_threading_layer_or_category_without_probing() -> ThreadingLayer | TheadingCategory:
```
Comment (flying-sheep) [inline src/fast_array_utils/numba.py]: this doesn’t need to exist, as not even the `cast` in it is necessary

File: tests/test_numba.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        with pytest.warns(UserWarning, match=warning):
+            assert wrapped() is expected
+    assert calls == [expected]
+
+
+def test_serial_fallback() -> None:
+    values = np.arange(10, dtype=np.float64)
+    wrapped = fa_numba.njit(_sum_prange)
+
+    with pytest.MonkeyPatch().context() as monkeypatch:
```
Comment (flying-sheep) [inline tests/test_numba.py]: why not just use the `monkeypatch` fixture like everywhere else?

**Review by flying-sheep — CHANGES_REQUESTED** [review]
Hi, thanks for this! I like the idea, with a subprocess even UB is unlikely to cause problems. Is the real fix tracked somewhere in the torch issue tracker or so?

I’m less of a fan of the gratuitous monkeypatching in the tests, but I get that avoiding it would mean re-architecting everything, e.g. by using a (data)class that can be initialized with all the variables that it by default gets from global constants and numba settings. No need to do that, I’ll do it when you’re done with the rest!

File: tests/test_numba.py (hunk 2 above)
Comment (JhonatanFelix) [reply tests/test_numba.py]: actually there's no need to a manual context here, I changed that test to just use the normal `monkeypatch`. Thank you

**Comment by JhonatanFelix** [comment]
> Hi, thanks for this! I like the idea, with a subprocess even UB is unlikely to cause problems. Is the real fix tracked somewhere in the torch issue tracker or so?
> 
> I’m less of a fan of the gratuitous monkeypatching in the tests, but I get that avoiding it would mean re-architecting everything, e.g. by using a (data)class that can be initialized with all the variables that it by default gets from global constants and numba settings. No need to do that, I’ll do it when you’re done with the rest!

Thanks!!! 
About the torch: I actually didn't find a `pytorch` issue that tracks the `scanpy/numba`crash itself. The closest upstream tracking I found was the problem of `pytorch` open `macOS/OpenMP` duplication issues, specially with the pytorch/pytorch#44282 and pytorch/pytorch#127973.

**Review by JhonatanFelix — COMMENTED** [review]
Made a wrong comment here, sorry!

File: src/fast_array_utils/numba.py [hunk 3]
```diff
@@ -217,18 +217,17 @@ def decorator(f: Callable[P, R], /) -> Callable[P, R]:
 
         @wraps(f)
         def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
+            msg = None
```
Comment (JhonatanFelix) [inline src/fast_array_utils/numba.py:114]: I think this change remove the early stop in the unsafe thread-pool path. So instead of changing directly to serial mode, the code keeps going and going and calls the parallel probe, which seems to break the expected behavior. I changed it back to the previous version ot the code to avoid that problem

File: src/fast_array_utils/numba.py [hunk 4]
```diff
@@ -220,17 +220,18 @@ def decorator(f: Callable[P, R], /) -> Callable[P, R]:
 
         @wraps(f)
         def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
-            msg = None
```
Comment (flying-sheep) [inline src/fast_array_utils/numba.py]: why did you undo my change?

File: src/fast_array_utils/numba.py (hunk 3 above)
Comment (flying-sheep) [reply src/fast_array_utils/numba.py:114]: I see! Please revert to my version with this change:

```diff
-if _needs_parallel_runtime_probe() and not _parallel_numba_runtime_is_safe():
+elif _needs_parallel_runtime_probe() and not _parallel_numba_runtime_is_safe():
```

Comment (JhonatanFelix) [reply src/fast_array_utils/numba.py:114]: Done!

**Comment by flying-sheep** [comment]
thank you!
